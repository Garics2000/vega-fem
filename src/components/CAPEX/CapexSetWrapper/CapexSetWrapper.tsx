import React, { useCallback, useEffect, useState } from 'react';
import { Button, Form, IconAdd, IconSelect, Text, TextField } from '@gpn-prototypes/vega-ui';

import CapexExpense from '../../../../types/CapexExpense';
import CapexExpenseSetGroup from '../../../../types/CapexExpenseSetGroup';
import CapexSet from '../../../../types/CapexSet';
import CapexSetGlobalValue from '../../../../types/CapexSetGlobalValue';
import keyGen from '../../../helpers/keyGenerator';
import { cnBlockWrapper } from '../../../styles/BlockWrapper/cn-block-wrapper';
import { cnVegaFormCustom } from '../../../styles/VegaFormCustom/cn-vega-form-custom';
import { FEMTable } from '../../FEMTable/FEMTable';

import { GroupWrapper } from './GroupWrapper';

import '../../../styles/BlockWrapper/BlockWrapper.css';
import '../../../styles/VegaFormCustom/VegaFormCustom.css';

interface CapexSetWrapperProps {
  capexSet: CapexSet;
  reservedValueSet: CapexSetGlobalValue;
  addCapexSetGroup: (capexSetGroup: CapexExpenseSetGroup) => void;
  addCapex: (capex: CapexExpense, group: CapexExpenseSetGroup) => void;
  updateCapexGlobalValue: (reserveValue: CapexSetGlobalValue) => void;
  updateCapexValue: (capex: CapexExpense, group: CapexExpenseSetGroup) => void;
}

export const CapexSetWrapper = ({
  capexSet,
  reservedValueSet,
  updateCapexGlobalValue,
  addCapexSetGroup,
  addCapex,
  updateCapexValue,
}: CapexSetWrapperProps) => {
  const [reserveValue, setReserveValue] = useState(reservedValueSet.value);
  const [reserveValueId, setReserveValueId] = useState(reservedValueSet.id);
  // const [years, setYears] = useState(capexSet.years);
  // const [yearStart,setYearStart]=useState(capexSet.yearStart);

  const [isAddingGroup, setIsAddingGroup] = useState(false);
  const [newGroupName, setNewGroupName] = useState('');
  const [groups, setGroups] = useState(
    (capexSet?.capexExpenseGroupList ?? []) as CapexExpenseSetGroup[],
  );

  useEffect(() => {
    if (reservedValueSet?.value) {
      setReserveValue(reservedValueSet.value);
    }
    if (reservedValueSet?.id) {
      setReserveValueId(reservedValueSet.id);
    }
    setGroups(capexSet?.capexExpenseGroupList ?? []);
  }, [reserveValueId, reservedValueSet, capexSet]);

  const toggleCapexSetGroup = (event: React.MouseEvent) => {
    event.preventDefault();
    setIsAddingGroup(!isAddingGroup);
    setNewGroupName('');
  };

  const requestGroupAdd = (groupName: string) => {
    addCapexSetGroup({
      caption: groupName,
    } as CapexExpenseSetGroup);
  };

  const addGroup = (event: any, groupName: string): void => {
    toggleCapexSetGroup(event);
    requestGroupAdd(groupName);
  };

  const onChangeTypoHandler = (e: any, setter: (value: any) => void) => {
    setter(e.e.target.value);
  };

  const requestSetGlobalValue = useCallback(() => {
    // TODO: запрос на обновление reservedValue
    console.log('reserveValue: ', reserveValue);
    updateCapexGlobalValue({
      id: reserveValueId,
      value: reserveValue,
    } as CapexSetGlobalValue);
  }, [updateCapexGlobalValue, reserveValue, reserveValueId]);

  return (
    <div className={cnBlockWrapper()}>
      <div className={cnBlockWrapper('title-wrapper')}>
        <div className={cnBlockWrapper('title')}>
          <Text as="span" size="xs" className={cnBlockWrapper('title-text')}>
            CAPEX
          </Text>
          <IconSelect size="xs" className={cnBlockWrapper('title-icon')} />
        </div>
      </div>
      <div className={cnBlockWrapper('content')}>
        {capexSet && !(Object.keys(capexSet).length === 0) ? (
          <>
            <Form
              className={cnVegaFormCustom()}
              onSubmit={(e: React.FormEvent) => {
                e.preventDefault();
              }}
            >
              <Form.Row gap="m" space="none" className={cnVegaFormCustom('content-body')}>
                <Form.Row gap="m" space="none" className={cnVegaFormCustom('form-row')}>
                  <Form.Field>
                    <Form.Label htmlFor="capexSetReservedValue">Величина резерва</Form.Label>
                    <TextField
                      id="capexSetReservedValue"
                      size="s"
                      width="full"
                      value={reserveValue?.toString()}
                      rightSide="%"
                      onBlur={() => requestSetGlobalValue()}
                      onChange={(e) => onChangeTypoHandler(e, setReserveValue)}
                    />
                  </Form.Field>
                </Form.Row>
                <Form.Row
                  col="1"
                  gap="none"
                  space="none"
                  className={cnVegaFormCustom('groups-row')}
                >
                  {(groups ?? []).length > 0 &&
                    groups.map((group, index) => (
                      <GroupWrapper
                        key={keyGen(index)}
                        group={group}
                        requestAddCapex={addCapex}
                        updateCapexValue={updateCapexValue}
                      />
                    ))}
                </Form.Row>
              </Form.Row>
              <Form.Row col="1" gap="none" space="none" className={cnVegaFormCustom('footer')}>
                {!isAddingGroup && (
                  <Button
                    size="s"
                    label="Добавить группу затрат"
                    iconLeft={IconAdd}
                    view="ghost"
                    onClick={(e) => toggleCapexSetGroup(e)}
                  />
                )}
                {isAddingGroup && (
                  <div>
                    <Text as="span" view="secondary" size="s">
                      Название группы затрат
                    </Text>
                    <Form.Row col="1" gap="none" className={cnVegaFormCustom('footer-text-field')}>
                      <Form.Field>
                        <TextField
                          size="s"
                          width="full"
                          id="capexSetGroupName"
                          type="text"
                          maxLength={150}
                          value={newGroupName}
                          onChange={(event: any) => setNewGroupName(event.e.target.value)}
                        />
                      </Form.Field>
                    </Form.Row>
                    <Form.Row className={cnVegaFormCustom('footer-action')}>
                      <Button
                        size="s"
                        label="Добавить группу"
                        view="ghost"
                        disabled={!newGroupName.length}
                        onClick={(e) => addGroup(e, newGroupName)}
                      />
                      <Button label="Отмена" view="clear" onClick={toggleCapexSetGroup} />
                    </Form.Row>
                  </div>
                )}
              </Form.Row>
            </Form>
            <FEMTable
              entity={capexSet}
              secondaryColumn="valueTotal"
              headers={['', 'Статья', 'Суммарное']}
            />
          </>
        ) : (
          <div />
          // <CapexSetPlaceholder text="Выберите один из макроэкономических сценариев" />
        )}
      </div>
    </div>
  );
};
