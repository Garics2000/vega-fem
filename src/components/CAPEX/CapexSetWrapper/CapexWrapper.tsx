import React, { useState } from 'react';
import { Form, TextField } from '@gpn-prototypes/vega-ui';

import CapexExpense from '../../../../types/CAPEX/CapexExpense';
import { cnVegaFormCustom } from '../../../styles/VegaFormCustom/cn-vega-form-custom';
import { cnGroupWrapper } from '../../MacroparameterSetWrapper/GroupWrapper/cn-group-wrapper';

import '../../../styles/BlockWrapper/BlockWrapper.css';
import '../../MacroparameterSetWrapper/GroupWrapper/GroupWrapper.css';

interface CapexWrapperProps {
  capex: CapexExpense;
  updateCapexValue: (capex: CapexExpense) => void;
}

export const CapexWrapper = ({ capex, updateCapexValue }: CapexWrapperProps) => {
  const [value, setValue] = useState(capex?.valueTotal);

  const editValues = (e: any): void => {
    setValue(e.e.target.value);
  };

  return (
    <Form.Row className={cnVegaFormCustom('form-row')} space="m">
      <Form.Field className={cnGroupWrapper('body-content')}>
        <Form.Label space="2xs">{capex.caption}</Form.Label>
        <TextField
          size="s"
          width="full"
          id={`capex_${capex?.name}`}
          placeholder="Значение"
          rightSide={capex?.unit}
          value={value?.toString()}
          onBlur={() => updateCapexValue({ ...capex, ...{ valueTotal: value } })}
          onChange={(e: any) => editValues(e)}
        />
      </Form.Field>
    </Form.Row>
  );
};
