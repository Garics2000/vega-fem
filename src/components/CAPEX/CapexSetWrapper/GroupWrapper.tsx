import React, { useState } from 'react';
import { IconArrowDown } from '@gpn-design/uikit/IconArrowDown';
import { Button, IconAdd, Text, useModal } from '@gpn-prototypes/vega-ui';

import Article from '../../../../types/Article';
import CapexExpenseSetGroup from '../../../../types/CAPEX/CapexExpenseSetGroup';
import keyGen from '../../../helpers/keyGenerator';
import { GroupPlaceholder } from '../../MacroparameterSetWrapper/GroupPlaceholder/GroupPlaceholder';
import { cnGroupWrapper } from '../../MacroparameterSetWrapper/GroupWrapper/cn-group-wrapper';
import { AddArticleModal } from '../../Shared/AddArticleModal/AddArticleModal';

import { CapexWrapper } from './CapexWrapper';

import '../../../styles/BlockWrapper/BlockWrapper.css';
import '../../MacroparameterSetWrapper/GroupWrapper/GroupWrapper.css';
// import {Article} from "../../Shared/AddArticleModal/AddArticleModal";
interface CapexSetWrapperGroupProps {
  group: CapexExpenseSetGroup;
  requestAddCapex: (capex: Article, group: CapexExpenseSetGroup) => void;
  updateCapexValue: (capex: Article, group: CapexExpenseSetGroup) => void;
}

export const GroupWrapper = ({
  group,
  requestAddCapex,
  updateCapexValue,
}: CapexSetWrapperGroupProps) => {
  const [capexes] = useState((group?.capexExpenseList ?? []) as Article[]);

  const [isCollapsed, setIsCollapsed] = useState(true);

  const { isOpen, close, open } = useModal();

  const openAddCapexModal = (): void => {
    setIsCollapsed(false);
    open();
  };

  const addCapexToGroup = (capex: Article): void => requestAddCapex(capex, group);

  const updateCapexValueWithGroup = (capex: Article): void => updateCapexValue(capex, group);

  return (
    <div className={cnGroupWrapper()}>
      <div className={cnGroupWrapper('header')}>
        <div
          className={cnGroupWrapper('header-name', { collapse: isCollapsed })}
          onClick={() => setIsCollapsed(!isCollapsed)}
          role="presentation"
        >
          <IconArrowDown size="xs" />
          <Text as="span" size="s">
            {group?.caption}
          </Text>
        </div>
        <div className={cnGroupWrapper('header-actions')}>
          <Button
            type="button"
            title="Добавить статью"
            onlyIcon
            iconLeft={IconAdd}
            size="s"
            view="clear"
            onClick={openAddCapexModal}
          />
        </div>
      </div>
      <div className={cnGroupWrapper('body', { hidden: isCollapsed })}>
        {capexes?.length === 0 && (
          <GroupPlaceholder text="Пока не добавлена ни одна статья" callback={openAddCapexModal} />
        )}
        {capexes?.length > 0 &&
          capexes.map((capex, index) => (
            <CapexWrapper
              key={keyGen(index)}
              capex={capex}
              updateCapexValue={updateCapexValueWithGroup}
            />
          ))}
      </div>
      <AddArticleModal
        isOpen={isOpen}
        close={close}
        article={{ caption: '', unit: '' } as Article}
        callback={addCapexToGroup}
      />
    </div>
  );
};
