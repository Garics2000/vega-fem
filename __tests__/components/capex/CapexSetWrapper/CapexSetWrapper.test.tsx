import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { CapexSetWrapper } from '../../../../src/components/CAPEX/CapexSetWrapper/CapexSetWrapper';
import CapexSet from '../../../../types/CAPEX/CapexSet';
import CapexSetGlobalValue from '../../../../types/CAPEX/CapexSetGlobalValue';

let fakeCapexSet: CapexSet;
let fakeReservedValue: CapexSetGlobalValue;

beforeEach(() => {
  fakeCapexSet = {
    years: 15,
    yearStart: 2020,
    capexGlobalValueList: [
      {
        id: '3',
        name: 'reserveValue',
        caption: 'Величина резерва',
        value: 38.0,
      },
    ],
    capexExpenseGroupList: [
      {
        id: '1',
        name: 'oneTimePaymentGroup',
        caption: 'Первоначальный взнос',
        valueTotal: 1000000.0,
        capexExpenseList: [
          {
            id: '1',
            name: 'oneTimePaymentValue',
            caption: 'Значение разового платежа',
            valueTotal: 1000000.0,
          },
        ],
      },
    ],
  };
  fakeReservedValue = {
    id: '3',
    name: 'reserveValue',
    value: 38.0,
  };
});

describe('<CapexSetWrapper/>', () => {
  test('рендерится без ошибок', () => {
    const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
      .create(
        <CapexSetWrapper
          capexSet={fakeCapexSet}
          reservedValueSet={fakeReservedValue}
          addCapexSetGroup={jest.fn()}
          addCapex={jest.fn()}
          updateCapexGlobalValue={jest.fn()}
          updateCapexValue={jest.fn()}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('срабатывают евенты на изменение глобального значения', () => {
    const fakeUpdateCapexGlobalValue = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        reservedValueSet={fakeReservedValue}
        addCapexSetGroup={jest.fn()}
        addCapex={jest.fn()}
        updateCapexGlobalValue={fakeUpdateCapexGlobalValue}
        updateCapexValue={jest.fn()}
      />,
    );

    const globalValueInput: HTMLElement = screen.getByDisplayValue(/38/);
    fireEvent.change(globalValueInput, { target: { value: '50' } });
    expect(globalValueInput).toHaveAttribute('value', '50');
    fireEvent.blur(globalValueInput);
    expect(fakeUpdateCapexGlobalValue).toBeCalledTimes(1);
  });

  test('срабатывают евенты на добавление группы', () => {
    const fakeAddGroup = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        reservedValueSet={fakeReservedValue}
        addCapexSetGroup={fakeAddGroup}
        addCapex={jest.fn()}
        updateCapexGlobalValue={jest.fn()}
        updateCapexValue={jest.fn()}
      />,
    );

    const addToggle: HTMLElement = screen.getByText(/добавить группу затрат/i);
    fireEvent.click(addToggle);
    const addGroupInput: HTMLElement = screen.getByPlaceholderText(
      'Введите название группы затрат',
    );
    const addGroupButton: HTMLElement = screen.getByText('Добавить группу');

    expect(addGroupInput).toBeInTheDocument();
    fireEvent.change(addGroupInput, { target: { value: 'new group' } });
    fireEvent.click(addGroupButton);
    expect(fakeAddGroup).toBeCalledTimes(1);
  });
  test('отмена добавления группы', () => {
    const fakeAddGroup = jest.fn();
    render(
      <CapexSetWrapper
        capexSet={fakeCapexSet}
        reservedValueSet={fakeReservedValue}
        addCapexSetGroup={fakeAddGroup}
        addCapex={jest.fn()}
        updateCapexGlobalValue={jest.fn()}
        updateCapexValue={jest.fn()}
      />,
    );

    const addToggle: HTMLElement = screen.getByText(/добавить группу затрат/i);
    fireEvent.click(addToggle);
    const addGroupInput: HTMLElement = screen.getByPlaceholderText(
      'Введите название группы затрат',
    );
    const addGroupCancelButton: HTMLElement = screen.getByText('Отмена');

    expect(addGroupInput).toBeInTheDocument();
    fireEvent.change(addGroupInput, { target: { value: 'new group' } });
    fireEvent.click(addGroupCancelButton);
    expect(fakeAddGroup).toBeCalledTimes(0);
  });
});
