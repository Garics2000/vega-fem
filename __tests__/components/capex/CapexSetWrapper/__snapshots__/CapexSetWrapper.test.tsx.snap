// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CapexSetWrapper/> рендерится без ошибок 1`] = `
<div
  className="BlockWrapper"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <span
        className="Text text text_size_xs text_view_primary"
      >
        CAPEX
      </span>
      <span
        className="Icon Icon_size_xs IconSelect"
      >
        <svg
          className="Icon-Svg"
          viewBox="0 0 12 12"
        >
          <path
            d="M6 8.5L10 4H2l4 4.5z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <form
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      onSubmit={[Function]}
    >
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <div
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          <div
            className={
              Object {
                "has": [Function],
                "is": [Function],
                "mix": [Function],
                "split": [Function],
                "state": [Function],
                "toString": [Function],
              }
            }
          >
            <label
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              htmlFor="capexSetReservedValue"
            >
              Величина резерва
            </label>
            <div
              className="TextField TextField_size_s TextField_view_default TextField_form_default TextField_width_full TextField_type_text TextField_withValue"
            >
              <input
                autoFocus={false}
                className="TextField-Input"
                id="capexSetReservedValue"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value="38"
              />
              <div
                className="TextField-Side TextField-Side_position_right TextField-Side_type_string"
              >
                %
              </div>
            </div>
          </div>
        </div>
        <div
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          <div
            className="GroupWrapper"
          >
            <div
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
                onClick={[Function]}
                role="presentation"
              >
                <span
                  className="Icon Icon_size_xs IconArrowDown"
                >
                  <svg
                    className="Icon-Svg"
                    viewBox="0 0 12 12"
                  >
                    <path
                      clipRule="evenodd"
                      d="M6.005 8.341L2.03 4.366l.707-.707 3.268 3.268 3.268-3.268.707.707-3.975 3.975z"
                      fillRule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  className="Text text text_size_s text_view_primary"
                >
                  Первоначальный взнос
                </span>
              </div>
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
              >
                <button
                  className="Button Button_size_s Button_view_clear Button_form_default Button_withIcon Button_onlyIcon"
                  onClick={[Function]}
                  title="Добавить статью"
                >
                  <span
                    className="Icon Icon_size_xs IconAdd Button-Icon"
                  >
                    <svg
                      className="Icon-Svg"
                      viewBox="0 0 12 12"
                    >
                      <path
                        d="M6 3H5v3H2v1h3v3h1V7h3V6H6V3z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
            <div
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              <div
                className={
                  Object {
                    "has": [Function],
                    "is": [Function],
                    "mix": [Function],
                    "split": [Function],
                    "state": [Function],
                    "toString": [Function],
                  }
                }
              >
                <div
                  className={
                    Object {
                      "has": [Function],
                      "is": [Function],
                      "mix": [Function],
                      "split": [Function],
                      "state": [Function],
                      "toString": [Function],
                    }
                  }
                >
                  <label
                    className={
                      Object {
                        "has": [Function],
                        "is": [Function],
                        "mix": [Function],
                        "split": [Function],
                        "state": [Function],
                        "toString": [Function],
                      }
                    }
                  >
                    Значение разового платежа
                  </label>
                  <div
                    className="TextField TextField_size_s TextField_view_default TextField_form_default TextField_width_full TextField_type_text TextField_withValue"
                  >
                    <input
                      autoFocus={false}
                      className="TextField-Input"
                      id="capex_oneTimePaymentValue"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="Значение"
                      type="text"
                      value="1000000"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <button
          className="Button Button_size_s Button_view_ghost Button_form_default Button_withIcon"
          onClick={[Function]}
        >
          <span
            className="Icon Icon_size_xs IconAdd Button-Icon Button-Icon_position_left"
          >
            <svg
              className="Icon-Svg"
              viewBox="0 0 12 12"
            >
              <path
                d="M6 3H5v3H2v1h3v3h1V7h3V6H6V3z"
              />
            </svg>
          </span>
          <span
            className="Button-Label"
          >
            Добавить группу затрат
          </span>
        </button>
      </div>
    </form>
    <div
      className="FEMTableWrapper"
    >
      <table
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <thead>
          <tr>
            <th>
              
            </th>
            <th>
              Статья
            </th>
            <th>
              Суммарное
            </th>
            <th>
              2020
            </th>
            <th>
              2021
            </th>
            <th>
              2022
            </th>
            <th>
              2023
            </th>
            <th>
              2024
            </th>
            <th>
              2025
            </th>
            <th>
              2026
            </th>
            <th>
              2027
            </th>
            <th>
              2028
            </th>
            <th>
              2029
            </th>
            <th>
              2030
            </th>
            <th>
              2031
            </th>
            <th>
              2032
            </th>
            <th>
              2033
            </th>
            <th>
              2034
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              title="Первоначальный взнос"
            >
              Первоначальный взнос
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              1000000
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            />
          </tr>
          <tr>
            <td />
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              title="Значение разового платежа"
            >
              Значение разового платежа
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
            >
              1000000
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
            <td
              className={
                Object {
                  "has": [Function],
                  "is": [Function],
                  "mix": [Function],
                  "split": [Function],
                  "state": [Function],
                  "toString": [Function],
                }
              }
              onBlur={[Function]}
              onDoubleClick={[Function]}
            >
              
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
